x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -548,61 +548,100 @@ public void checkPrivileges(
+                ? ResourcePrivilegesMap.builder()
+                : null;
             for (RoleDescriptor.ApplicationResourcePrivileges p : privilegesToCheck.application()) {
                 if (applicationName.equals(p.getApplication())) {
-                    ResourcePrivilegesMap appPrivsByResourceMap = userRole.checkApplicationResourcePrivileges(
+                    boolean privilegesGranted = userRole.checkApplicationResourcePrivileges(
                         applicationName,
                         Sets.newHashSet(p.getResources()),
                         Sets.newHashSet(p.getPrivileges()),
-                        applicationPrivileges
+                        applicationPrivileges,
+                        resourcePrivilegesMapBuilder
                     );
-                    builder.addResourcePrivilegesMap(appPrivsByResourceMap);
+                    allMatch = allMatch && privilegesGranted;
+                    if (false == privilegesToCheck.runDetailedCheck() && false == privilegesGranted) {
+                        maybeCachingListener.onResponse(PrivilegesCheckResult.SOME_CHECKS_FAILURE_NO_DETAILS);
+                        return;
+                    }
                 }
             }
-            ResourcePrivilegesMap resourcePrivsForApplication = builder.build();
-            allMatch = allMatch && resourcePrivsForApplication.allAllowed();
-            privilegesByApplication.put(applicationName, resourcePrivsForApplication.getResourceToResourcePrivileges().values());
+            if (resourcePrivilegesMapBuilder != null) {
+                privilegesByApplication.put(
+                    applicationName,
+                    resourcePrivilegesMapBuilder.build().getResourceToResourcePrivileges().values()
+                );
+            }
         }
 
-        final PrivilegesCheckResult privilegesCheckResult = new PrivilegesCheckResult(
-            allMatch,
-            cluster,
-            allIndices.getResourceToResourcePrivileges(),
-            privilegesByApplication
-        );
-        ActionListener.runBefore(listener, () -> {
-            if (userRole instanceof SimpleRole simpleRole) {
-                simpleRole.cacheHasPrivileges(settings, privilegesToCheck, privilegesCheckResult);
-            }
-        }).onResponse(privilegesCheckResult);
+        if (privilegesToCheck.runDetailedCheck()) {
+            maybeCachingListener.onResponse(
+                new PrivilegesCheckResult(
+                    allMatch,
+                    new PrivilegesCheckResult.Details(
+                        clusterPrivilegesCheckResults,
+                        combineIndicesResourcePrivileges.build().getResourceToResourcePrivileges(),
+                        privilegesByApplication
+                    )
+                )
+            );
+        } else {
+            maybeCachingListener.onResponse(PrivilegesCheckResult.ALL_CHECKS_SUCCESS_NO_DETAILS);
```

Comment:
Should we assert `allMatch` is indeed `true`?

Reply:
Yep, I made it so.